src/translate-virtual-address.h

```diff
@@ -148,26 +158,40 @@ static NO_INLINE bool translate_virtual_address(STATE_ACCESS &a, uint64_t *ppadd
     // Hence, the use of arithmetic shifts here
     int vaddr_bits = XLEN - (LOG2_PAGE_SIZE + levels * LOG2_VPN_SIZE);
     if (unlikely((static_cast<int64_t>(vaddr << vaddr_bits) >> vaddr_bits) != static_cast<int64_t>(vaddr))) {
-        return false;
+        *ppaddr = vaddr;
+        return to_mcause_code<TRANSLATION_MODE, ACCESS_TYPE>();
     }
 
     // Initialize pte_addr with the base address for the root page table
-    uint64_t pte_addr = (satp & SATP_PPN_MASK) << LOG2_PAGE_SIZE;
+    uint64_t pte_addr = (atp & SATP_PPN_MASK) << LOG2_PAGE_SIZE;
     for (int i = 0; i < levels; i++) {
+        int vpn_bits = LOG2_VPN_SIZE;
+        if (i == 0) vpn_bits += widenbits;
+        uint64_t vpn_mask = (1 << vpn_bits) - 1;
+
         // Mask out VPN[levels-i-1]
-        int vaddr_shift = LOG2_PAGE_SIZE + LOG2_VPN_SIZE * (levels - 1 - i);
-        uint64_t vpn = (vaddr >> vaddr_shift) & VPN_MASK;
+        int vaddr_shift = LOG2_PAGE_SIZE + vpn_bits * (levels - 1 - i);
+        uint64_t vpn = (vaddr >> vaddr_shift) & vpn_mask;
         // Add offset to find physical address of page table entry
         pte_addr += vpn << LOG2_PTE_SIZE; //??D we can probably save this shift here
+        if constexpr (TRANSLATION_MODE == TRANSLATION_VS) {
```

Comment:
This code could benefit from a comment of what is going on here, took me a while to understand. Preferably quoting the spec if possible.